### PR TITLE
Support passing boolean flags to kubectl plugins

### LINKED ISSUE #53640: Not complete support for flags of Boolean type in kubectl plugins
When invoking any plugin by kubectl, it exports the options specified in plugin YAML file as env vars. If there are any options that need to be designed as Boolean type, for example, if you want to have a Boolean type option "create-instance" then the plugin should always be invoked as follows:
```
kubectl plugin <plugin_name> --create-instance=true
``` 
instead of just 
```
kubectl plugin <plugin_name> --create-instance
```
You need to explicitly specify the value otherwise the env var values get messed up especially when you have multiple various type options.

### PR DESCRIPTION
**Pull request by carolynvs**
**What this PR does / why we need it**:
This allows kubectl plugins to define boolean flags. Boolean flags are set to true when the flag is present and no value is specified. --flag is equivalent to --flag=true and is useful for commands such as `kubectl plugin svcat --version`.

To preserve backwards compatibility with existing plugins, all flags by default are string flags, but now a new field has been introduced to the flag definition, type, which allows specifying the type of pflag to use. Currently only string and bool are supported, but this allows for other types to be added in the future.

Bool flags support a default value, and when defValue is set, kubectl validates that it can be parsed as a boolean when validating the plugin. Acceptable values can be found at https://golang.org/pkg/strconv/#ParseBool.

**Example**
```yaml
name: "svcat"
shortDesc: "Service Catalog CLI"
command: "./svcat"
flags:
  - name: version
    desc: Display the version of the plugin
    type: bool
```

**Which issue(s) this PR fixes**
Fixes #53640

**Special notes for your reviewer**:
None

**Release note**:
```release-note
Support passing boolean flags to kubectl plugins
```

/sig cli

### CODE REVIEW

**Comment by ericchiang** [comment]
/ok-to-test

cc @fabianofranz are you a good person to review? Maybe someone else from sig/cli?

**Comment by ericchiang** [comment]
cc @kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/plugin.go
Comment (dixudx) [inline pkg/kubectl/cmd/plugin.go]: IMO, better make it as default.

File: pkg/kubectl/plugins/plugins.go
Comment (dixudx) [inline pkg/kubectl/plugins/plugins.go]: What if `DefValue` is empty for boolean type?

File: pkg/kubectl/plugins/plugins.go
Comment (dixudx) [inline pkg/kubectl/plugins/plugins.go]: I prefer to move this into function `Validate`. No need to expose here.

File: pkg/kubectl/plugins/plugins.go
Comment (dixudx) [inline pkg/kubectl/plugins/plugins.go]: We should make this backwards compatible. If the flag type is missing or empty string, we should take it as string flag by default.

File: pkg/kubectl/plugins/plugins.go
Comment (dixudx) [inline pkg/kubectl/plugins/plugins.go]: To make it backwards compatible, this new-added field should be optional.

Should be `json:"type, omitempty"` here.

File: pkg/kubectl/plugins/plugins.go
Comment (carolynvs) [reply pkg/kubectl/plugins/plugins.go]: You're right, I'll update the patch so that it defaults to `false` when not specified for boolean flags.

**Comment by carolynvs** [comment]
Thanks for the feedback, @dixudx! I've updated the PR to incorporate your suggestions.

**Review by dixudx — COMMENTED** [review]
Only a small nit. LGTM mostly.

**Comment by dixudx** [comment]
@carolynvs Please also squash your commits. Then lgtm. Thanks.

**Comment by carolynvs** [comment]
@dixudx I've updated the error message indicating that strings are the default type and squashed my commits.

**Comment by dixudx** [comment]
/lgtm
@carolynvs Good job! Thanks for your work.

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by ericchiang** [comment]
test failure looks unrelated

/test pull-kubernetes-bazel-test

**Comment by dixudx** [comment]
Ping @kubernetes/sig-cli-maintainers for approval. Thanks.

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/58768)

**Comment by carolynvs** [comment]
None of the failing tests are related to my PR. I'm not quite sure what I should do next? 🤷‍♀️

**Comment by adohe** [comment]
This overall looks good to me, I would like to get this merged asap. Then I will open a new issue to put things which we need to do before we can move kubectl plugin to beta.

**Review by adohe-zz — APPROVED** [review]

**Comment by adohe** [comment]
Seems milestone freeze, could be get merged after the 1.10 release.

**Comment by liggitt** [comment]
> This overall looks good to me, I would like to get this merged asap. Then I will open a new issue to put things which we need to do before we can move kubectl plugin to beta.

The current plugin design led to the necessity for this pull. Before going further down the road of requiring plugins to describe all their flags to kubectl, it’s worth discussing whether that is the correct approach at all. That approach stunts plugins by requiring all their flags to be parseable by kubectl. I’d be strongly in favor of just passing through all envvars, flags and input/output streams and letting plugins run as if you had just invoked them directly.

**Comment by deads2k** [comment]
> @liggitt I could not agree more that having kubectl parse the Service Catalog's plugin flags makes things much more difficult for us, and it's not clear what is being gained. Our plugin is a binary that uses cobra, so for us it's a pointless exercise to generate the plugin yaml, use viper to rebind the envvars back to the original flags, etc. If there is a conversation about this going on elsewhere, I'd love to provide feedback!

We're mostly busy closing down 1.10, but I think that we should reconsider how plugins for `kubectl` work.  Straight pass-through with execve avoids unnecessary meta, allows easier testing, and makes the plugged in binaries make sense both with and without `kubectl`.

@soltysh

**Comment by soltysh** [comment]
@adohe @carolynvs let's discuss the idea @deads2k voiced out during our next sig-cli call. That's basically the direction I'd like us take before we promote this to beta. 
/lgtm cancel

**Comment by liggitt** [comment]
c.f. sig-cli meeting info (https://github.com/kubernetes/community/tree/master/sig-cli) and agenda (https://docs.google.com/document/d/1r0YElcXt6G5mOWxwZiXgGu_X6he3F--wKwg-9UBc29I/edit#)

**Comment by soltysh** [comment]
@carolynvs have a look at https://github.com/kubernetes/community/pull/2437 and https://github.com/kubernetes/kubernetes/pull/66876 which should address all your problems. This new mechanism should unblock all the problems you were struggling with  before. Ping me if you have any questions/problems.
/close

**Comment by carolynvs** [comment]
@soltysh This is great, thanks! Looking forward to the next release of k8s. I'm working with SIG Service Catalog to get svcat updated to work with the new model. 👍

**Pull request closed without merging**
